From e911527503f59411f7de7cefe70885b7f4267f37 Mon Sep 17 00:00:00 2001
From: Gleb Popov <arrowd@FreeBSD.org>
Date: Thu, 2 Oct 2025 13:14:54 +0300
Subject: [PATCH 364/553] QNativeSocketEngine::read: Only treat readBytes==0 as
 a EOF condition with stream sockets

This change also covers SOCK_SEQPACKET type of sockets, which isn't great,
but allows to avoid an endless polling loop when waiting for read on
a disconnected socket.

Change-Id: I6117b97c2c9b775e9e6fe9757bf4374365e3227d
Reviewed-by: Thiago Macieira <thiago.macieira@intel.com>
(cherry picked from commit 8e06365bee31ef5a1980b7b24d1564e3d769cfc7)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/network/socket/qnativesocketengine.cpp | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/src/network/socket/qnativesocketengine.cpp b/src/network/socket/qnativesocketengine.cpp
index 9abafc03ad9..7173461e65e 100644
--- a/src/network/socket/qnativesocketengine.cpp
+++ b/src/network/socket/qnativesocketengine.cpp
@@ -905,12 +905,11 @@ qint64 QNativeSocketEngine::read(char *data, qint64 maxSize)
 
     qint64 readBytes = d->nativeRead(data, maxSize);
 
-    // Handle remote close
-    if (readBytes == 0 && (d->socketType == QAbstractSocket::TcpSocket
-#ifndef QT_NO_SCTP
-        || d->socketType == QAbstractSocket::SctpSocket
-#endif
-        )) {
+    // Handle remote close.
+    // Non-datagram socket types signal the EOF state with a zero read.
+    // Note that it is perfectly fine to have a 0-byte message with datagram
+    // sockets (SOCK_DGRAM or SOCK_SEQPACKET).
+    if (readBytes == 0 && d->socketType != QAbstractSocket::UdpSocket) {
         d->setError(QAbstractSocket::RemoteHostClosedError,
                     QNativeSocketEnginePrivate::RemoteHostClosedErrorString);
         close();
-- 
2.51.2

